### PR TITLE
Exclude local plugins from bundle

### PR DESCRIPTION
**Pull request by shalvah**
Closes: #7908

Auto-excludes local plugins in `serverless.yml` from the serverless bundle. Supports files and folders, eg:

```yaml
- plugins:
  # Path to a folder
  - ./custom-plugins/Plugin1
  # Same, but with trailing slash
  - ./custom-plugins/Plugin2/
  # Path to a specific file
  - ./custom-plugins/Plugin3/index.js
  # Path to a specific file, but without extension
  - ./custom-plugins/Plugin3/index
```

## How does this work?
- When loading local plugins, the PluginManager records the CJS-resolved path as well as the original path as written in `serverless.yml`. Note that the CJS-resolved path always points to an entry file—usually `index.js, but not always.
- During packaging, `getLocalPluginsPaths()` compares each entry file to the original specified path and returns an absolute path pointing to the file or folder to exclude.

----

PS: I'm a big fan and looking to contribute. Would like to implement https://github.com/serverless/serverless/issues/9116, but decided to start with this, since this is one of its dependencies. Hopefully we can get this in soon🤞

### CODE REVIEW

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:148]: I don't think we need to change this function (and convert it to the method).

It'll be better, to register eventual local plugins in this context: https://github.com/serverless/serverless/blob/ce66591a688ca734eb34eff929163134d1d21ae8/lib/classes/PluginManager.js#L180-L212
Note that there `isLocalPlugin` variable answers wether it's a path we should consider for packaging exclusion

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:263]: Let's not create such methods (there's no real value in creating alias for `this.localPluginsPaths.push`

File: lib/classes/PluginManager.js
Comment (medikoo) [reply lib/classes/PluginManager.js:148]: Also what's important:
- We should only group there paths to local plugins (so those which are referenced by relative paths) and not all plugins 
- We should detect whether path targets a file or directory, and in case of a directory ensure all _directory_ is excluded. This is slightly tricky to accomplish. Logic I propose is to to compare against CJS resolved path with something as:

```javascript
if (path.extname(pluginPath) === path.extname(pluginResolvedPath)) {
  // assume file, add `pluginResolvedPath` to `localPluginsPaths`
}
if (pluginResolvedPath.slice(0, -path.extname(pluginResolvedPath).length)).endsWith(path.sep + path.basename(pluginPath)) {
  // assume file, add `pluginResolvedPath` to `localPluginsPaths`
}
if (path.dirname(pluginResolvedPath).endsWith(path.sep + pluginPath)) {
  // assume dir, add `path.dirname(pluginResolvedPath)` to `localPluginsPaths`
}
```

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:226]: I think we wont' need that method, please see my other comment

File: lib/plugins/package/lib/packageService.js
Comment (medikoo) [inline lib/plugins/package/lib/packageService.js:39]: This part should not be changed afaik.

it's just that `localPathExcludes` should be enriched with paths that we'd group at `this.serverless.pluginManager.localPluginsPaths`

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @shalvah for the initiative!

It's great to see someone trying to bring this down. Still please see my comments, I believe we need a slightly different approach. If something is not clear, let me know

File: lib/classes/PluginManager.js
Comment (shalvah) [reply lib/classes/PluginManager.js:148]: > We should detect whether path targets a file or directory, and in case of a directory ensure all directory is excluded

@medikoo this approach already does that. Did you check out the `getLocalPluginsPaths()` method and the tests? 

https://github.com/serverless/serverless/pull/10259/files#diff-888782bdfff1d922f13ff8f30784e31610be569c7fb93a8a3bd2611c1808db11R235-R253

File: lib/plugins/package/lib/packageService.js
Comment (shalvah) [reply lib/plugins/package/lib/packageService.js:39]: @medikoo Perhaps, but I don't think there's any harm. We're still achieving the same thing, yes? "Give me all the paths for local plugins". This way, it's a straightforward request to the plugin manager. Does this affect anything? Am I missing something?

File: lib/classes/PluginManager.js
Comment (medikoo) [reply lib/classes/PluginManager.js:148]: @shalvah thanks, indeed it does an attempt on that. That's great.

Still I think we need to take into account only relative paths (those that starts with `./` and not all paths), and also we do not need to do `fs.realpathSync` resolution, I think it's more accurate to work with the outcome of `cjsResolve` (it returns fully resolved real paths)

Comment (shalvah) [reply lib/classes/PluginManager.js:148]: > I don't think we need to change this function (and convert it to the method).


Fair, so you're saying we should move the `this.localPluginsPaths` to the calling method? That can work, but we'll need to run `cjsResolve` again to get the correct entry file path so we can figure out whether it's a file or directory. I'm not sure how expensive `cjsResolve(0` is.

Comment (shalvah) [reply lib/classes/PluginManager.js:148]: > Note that there isLocalPlugin variable 

Ah, I just noticed this. But is the variable definition correct? `name.startsWith('./')`. What if a user specifies a path as '../`? Is that supported.

File: lib/plugins/package/lib/packageService.js
Comment (medikoo) [reply lib/plugins/package/lib/packageService.js:39]: The problem to me, is that we touch here logic we should keep. We also need to be careful with changes to not break anything, so instead of refactoring reshaping existing parts, we should rather leave them intact, and just add 2 cents which resolves those few more paths we need to cover in exclusion.

So my point is that the resolution of `const pluginsLocalPath` should remain the same (be left untouched), we still need that path in the form it was resolved. Let's just add logic which adds those extra paths to `localPathExcludes`, that's it.

Also, it's good to keep old handling of local path plugins with new handling as separate, as it might be that we will want to drop support for the old way, if we mix it, it'll be harder

Comment (shalvah) [reply lib/plugins/package/lib/packageService.js:39]: Another advantage of doing it this way: when we remove support for `.serverless_plugins` and `plugins.localPath`, as I've seen is planned, the only change will be to not add it to the `localPluginsPaths` array in the plugin manager. The packager does not have to know or care about `.serverless_plugins`, and its code will not change,

File: lib/classes/PluginManager.js
Comment (medikoo) [reply lib/classes/PluginManager.js:148]: > Ah, I just noticed this. But is the variable definition correct? name.startsWith('./'). What if a user specifies a path as '../`? Is that supported?

This would mean that user wants to reach out beyond service folder, we should not support that way (it's safe to ignore)

File: lib/plugins/package/lib/packageService.js
Comment (medikoo) [reply lib/plugins/package/lib/packageService.js:39]: > the only change will be to not add it to the localPluginsPaths array in the plugin manager. 

Why do you assume so? I think we need this collection to collect there all local plugins referenced by relative paths.

**Comment by shalvah** [comment]
@medikoo I'm very new to the internals of this project, so I'm still a bit confused. Trying to separate stylistic changes from logic changes. Thus far, I've got:
1. Get rid of extra method `getLocalPluginsPaths` (stylistic)
2. Keep ``requireServicePlugin` as a local function (stylistic)
3. Use `isLocalVariable`  to only exclude local plugins (logic), but this will require `cjsResolve()` ing again.

You've also mentioned the file vs directory thing (logic), but I think the current approach already handles this robustly. 

I've made some changes. I implemented (1) and achieved (3), but not in the exact way. Please take a look and let me know if there are still issues

File: lib/classes/PluginManager.js
Comment (shalvah) [reply lib/classes/PluginManager.js:148]: > also we do not need to do fs.realpathSync resolution, I think it's more accurate to work with the outcome of cjsResolve (it returns fully resolved real paths)

They're complementary, not replacements.

For instance, imagine a folder `myplugin` with a `main` entry in `package.json` pointing to `src/main.js`. The user can pass in `./myplugin`, which will work with CJS (resolved to `/path/myplugin/src/main.js`). By checking the realpath of `/myplugin`, we'll get `/path/myplugin`, which we can then compare with the CJS-resolved path for a match. We could probably figure it out somehow using the specified relative path, but it's simpler to just check the realpath.

File: lib/plugins/package/lib/packageService.js
Comment (shalvah) [reply lib/plugins/package/lib/packageService.js:39]: Hmmm...I don't really understand.

But I see your point about not refactoring existing code (though I don't entirely agree😀). I'll make this change.

**Comment by shalvah** [comment]
Sorry for the back-and-forth @medikoo.😅 I think I'm done now. Let me know if there's anything outstanding.

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:226]: I would name `entryFileRealPath` as `resolvedPath` and `originalPath` as `inputPath`

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:228]: While it's a corner case, user may use folder that ends with `.js`.

Solid approach would be to probably compare something as:

```javascript
if (path.resolve(serviceDir, inputPath) === realPath) // assume realPath
```

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:237]: I'm still not convinced we need another fs check here. Wouldn't it be better to go as follows:

```javascript
const absoluteInputPath = path.resolve(serviceDir, inputPath);
if ((absoluteInputPath + path.extname(realPath)) === realPath) // assume realPath (file)
if (realPath.startsWith(absoluteInputPath + path.sep)) // assume absoluteInputPath (dir)
```

File: lib/plugins/package/lib/packageService.js
Comment (medikoo) [inline lib/plugins/package/lib/packageService.js:47]: As mentioned in previous comments, it'll be great to resolve patterns on spot, and then here simply:

```javascript
const localPathExcludes = Array.from(this.serverless.pluginManager.localPluginPathPatterns)
if (pluginsLocalPath) localPathExcludes .push(pluginsLocalPath)
```

**Review by medikoo — COMMENTED** [review]
Thank you @shalvah I think we're on the right track :) Please see my comments

**Comment by shalvah** [comment]
Good points @medikoo . I'll make the changes, and look into whether I can do away with `fs.realpath`.

File: test/unit/lib/classes/PluginManager.test.js
Comment (shalvah) [inline test/unit/lib/classes/PluginManager.test.js:1847]: This test is a copy of the one above it; removed it.

**Comment by shalvah** [comment]
Simplified things now @medikoo. Thanks. One thing, though: I kept the computation in the extra method call, because I don't think there's a point to doing the path resolution when it may not be needed. For example, a user might want to test their offline function or some other thing—no point in generating paths to exclude when there's not going to be any packaging.

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:230]: I think resolved pattern should be in form of a path that's relative to service path (?), while here absolute path is returned

File: test/unit/lib/classes/PluginManager.test.js
Comment (medikoo) [inline test/unit/lib/classes/PluginManager.test.js:1848]: We have a policy to configure all new tests with [`runServerless`](https://github.com/serverless/test/blob/master/docs/run-serverless.md). We have already plenty of such tests configured.

I think here it'll be good to rely on [`packaging`](https://github.com/serverless/serverless/tree/master/test/fixtures/programmatic/packaging) fixture, and extend configuration with local plugins that resemble direct file and directory, and confirm that in generated artifact those files are not included

**Review by medikoo — COMMENTED** [review]
Thank you @shalvah, It looks way better!

File: test/unit/lib/classes/PluginManager.test.js
Comment (shalvah) [reply test/unit/lib/classes/PluginManager.test.js:1848]: You mean an integration test? Yes, I agree. That was my first choice, but I couldn't locate any existing tests that checked packaging that I could copy from. I'll try again.

Comment (medikoo) [reply test/unit/lib/classes/PluginManager.test.js:1848]: No, I mean unit test (in sense that we can run it offline without dependency on AWS or any other remote instance).

Ideally if test is constructed with `runServerless` util, and then contents of result artifact is confirmed, in similar way as it is done in our old packaging tests: https://github.com/serverless/serverless/blob/ce66591a688ca734eb34eff929163134d1d21ae8/test/integrationPackage/lambda-files.tests.js#L52 (they should be at some point moved to regular unit tests and refactored to rely on `runServerless`)

Comment (shalvah) [reply test/unit/lib/classes/PluginManager.test.js:1848]: Yeah,I shoulda called them "eature tests" instead.

Added! Thanks for the guidance.

**Comment by shalvah** [comment]
@medikoo anything else blocking this PR?

**Comment by medikoo** [comment]
@shalvah sorry, I wasn't sure it's ready for re-review (normally it's best if after applying the changes you re-request review). Looking into it now

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:240]: I think we should return `null` and add filter to not return this value in collection

File: test/unit/lib/classes/PluginManager.test.js
Comment (medikoo) [inline test/unit/lib/classes/PluginManager.test.js:1848]: All new tests should be configured with `runServerless`, so if that case is converted by added `runServerless` run, then simply let's remove this test. Otherwise it'll be great to ensure it's covered with `runServerless` based test

**Review by medikoo — COMMENTED** [review]
Thank you @shalvah, we're very close! Please see my comments

File: lib/classes/PluginManager.js
Comment (shalvah) [reply lib/classes/PluginManager.js]: Gotcha.

**Review by medikoo — CHANGES_REQUESTED** [review]
@shalvah there are CI fails, If I see correctly some changes do not match prettier formatting

**Comment by shalvah** [comment]
My bad, fixed.

Not sure why the linting is failing though.

```
Run npm run lint:updated

> serverless@2.70.0 lint:updated
> pipe-git-updated --ext=js -- eslint


/home/runner/work/serverless/serverless/test/unit/lib/plugins/package/lib/packageService.test.js
  242:28  error  'mockedDescribeStacksResponse' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)
```

The variable is defined, and that isn't even line 242 in my code. Also, it passes on my machine.

**Comment by shalvah** [comment]
Oops, clicked on Close by mistake.

@medikoo any idea what's causing the linter to fail? It looks like it's linting some other branch or commit.

**Review by medikoo — COMMENTED** [review]
@shalvah CI tests branch of this PR merged with the current state of master (as it's what will be the final effect after we merge it to master), and such setup produces those fails.

It looks that in a meantime there were some changes made in this test file, and you need to rebase against master to see all of them and replicate the issue on your side

**Comment by shalvah** [comment]
Fixed now. Thanks for the help @medikoo . Ready to merge!

File: test/unit/lib/classes/PluginManager.test.js
Comment (medikoo) [inline test/unit/lib/classes/PluginManager.test.js:1840]: Why those tests were removed?

File: lib/classes/PluginManager.js
Comment (medikoo) [inline lib/classes/PluginManager.js:243]: We can also simplify it to `.filter(Boolean)`

**Review by medikoo — COMMENTED** [review]
Thanks @shalvah, that looks great. I have just one concert left towards removed tests. Let me know

File: lib/classes/PluginManager.js
Comment (shalvah) [reply lib/classes/PluginManager.js:243]: I was going to do that, but I felt you'd favour the explicitness. No harm leaving it this way, right?

File: test/unit/lib/classes/PluginManager.test.js
Comment (shalvah) [reply test/unit/lib/classes/PluginManager.test.js:1840]: It's a duplicate of the one above it.

File: lib/classes/PluginManager.js
Comment (medikoo) [reply lib/classes/PluginManager.js:243]: Yes, it's fine to leave it

File: test/unit/lib/classes/PluginManager.test.js
Comment (medikoo) [reply test/unit/lib/classes/PluginManager.test.js:1840]: Indeed, thanks for cleaning that up

**Review by medikoo — APPROVED** [review]
Thank you @shalvah !

**Comment by shalvah** [comment]
Thanks @medikoo  I'm ready to get started on #7909 (and then #9116). I'm just wondering about releases...the master branch is fr the current v2, yes? Do I need to PR my changes to the v3 branch as well?

**Comment by medikoo** [comment]
@shalvah please hold with #7909 as this is not approved internally. Technically just packages with `help wanted` label are ok to be taken.

Otherwise just work against `master` needs to be done (anything that's merged to master will be ported by us to `v3`, you don't need to take care about that).

Still if you're about to take task that may take more than few days it might be wiser to hold, as we plan to release v3 shortly, and if PR won't land in `master` by the time v3 is released, it'll have to be rebased against v3 release, which is likely to introduce a lot of conflicts (we plan thorough file rename across repository to unify convention, and to clean all legacy logs)

**Comment by mnapoli** [comment]
This change is breaking Bref (noticed that playing with the v3 beta): indeed the plugin file is distributed in the runtime package. And Bref being distributed via Composer (not NPM), the inclusion is done via a relative path (`./vendor/bref/bref`). That means that the whole package is excluded from the deployment.

Unfortunately I think we may have to revert this?

**Comment by medikoo** [comment]
@mnapoli if it's breaking, indeed we should revert.

Technically this PR was about ensuring consistency in how we handle plugins (by default they're never packaged for lambdas), and this change ideally should have landed together with support for `./*` local paths, but it was missed back then.

**Pull request merged**
